aten/src/ATen/native/mps/TensorFactory.cpp

```diff
@@ -0,0 +1,218 @@
+//  Copyright © 2022 Apple Inc.
+
+#include <ATen/ATen.h>
+#include <ATen/Tensor.h>
+#include <ATen/Utils.h>
+#include <torch/library.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/mps/Copy.h>
+#include <ATen/native/mps/TensorFactory.h>
+namespace at { namespace native {
+
+static inline void maybe_resize_storage_mps(TensorImpl* self, uint64_t new_size) {
+  if (new_size == 0) {
+    return;
+  }
+
+  auto storage = self->storage().unsafeGetStorageImpl();
+  if (!storage) {
+    TORCH_CHECK(false, "Tensor: invalid null storage");
+  }
+  uint64_t new_size_bytes = (new_size + self->storage_offset()) * self->dtype().itemsize();
+  if (new_size_bytes > self->storage().nbytes()) {
+    if (new_size_bytes == 0) {
+      storage->set_data_ptr_noswap(at::DataPtr(nullptr, at::Device(at::DeviceType::MPS, 0)));
+      storage->set_nbytes(0);
+    } else {
+      at::DataPtr new_data = storage->allocator()->allocate(new_size_bytes);
+      size_t copy_capacity = std::min<size_t>(new_size_bytes, storage->nbytes());
+      if (storage->data() && copy_capacity > 0) {
+        at::native::mps::copy_blit_mps(new_data.get(), storage->data(), copy_capacity);
+      }
+      // Destructively overwrite data_ptr
+      storage->set_data_ptr_noswap(std::move(new_data));
+      storage->set_nbytes(new_size_bytes);
+    }
+  }
+}
+
+inline TensorImpl* resize_impl_mps_(
+    TensorImpl* self,
+    IntArrayRef size,
+    c10::optional<IntArrayRef> stride,
+    bool device_guard = true) {
+  if (self->sizes() == size && (!stride || self->strides() == stride)) {
+    return self;
+  }
+
+  int64_t storage_size = 1;
+  if (stride) {
+    self->set_sizes_and_strides(size, *stride);
+    // NB: storage size can be different from numel.
+    storage_size = storage_size_for(size, *stride);
+  } else {
+    self->set_sizes_contiguous(size);
+    storage_size = self->numel();
+  }
+  maybe_resize_storage_mps(self, storage_size);
+
+  return self;
+}
+
+Tensor empty_mps(
+    IntArrayRef size,
+    c10::optional<ScalarType> dtype_opt,
+    c10::optional<Layout> layout_opt,
+    c10::optional<Device> device_opt,
+    c10::optional<bool> pin_memory_opt,
+    c10::optional<c10::MemoryFormat> memory_format_opt) {
+
+  return at::detail::empty_mps(size, dtype_opt, layout_opt, device_opt, pin_memory_opt, memory_format_opt);
+}
+
+Tensor empty_strided_mps(
+    IntArrayRef size,
+    IntArrayRef stride,
+    c10::optional<ScalarType> dtype_opt,
+    c10::optional<Layout> layout_opt,
+    c10::optional<Device> device_opt,
+    c10::optional<bool> pin_memory_opt) {
+  check_size_nonnegative(size);
+  // empty memory formatempty
+  auto t = at::native::empty_mps(
```

Comment:
Is there a reason that this can't be implemented using `empty_strided_generic()`, similar to how `empty_strided_cuda()` is implemented? (Where the extra unique cuda info is basically initting the cuda context, and specifying the cuda allocator + cuda dispatch key)

Reply:
Can I add a TODO to be followed up in future PR.? @bdhirsh